### ISSUE TITLE
QuerySet operations returns Sequence instead of List

### ISSUE THREAD

**Issue by dudil** [collerek/ormar#120]
**Describe the bug**
The QuerySet operations return value is *Sequence["Model"]* even though on the DocString it is stated as *List["Model"]*
It is causing for lint errors as mismatch types when trying to retrieve the returned object as List.

for example:
```python
rows: List[User] = await User.objects.all()
```
Will cause a lint error on mismatch type.

Looking into Ormar source code it seems the first time where this is starting to get confused is at:
ormar.models.mixins.merge_mixin.py ==> *merge_instances_list* class method

from source-code:
``` python
@classmethod
    def merge_instances_list(cls, result_rows: Sequence["Model"]) -> Sequence["Model"]:
        """
        Merges a list of models into list of unique models.

        Models can duplicate during joins when parent model has multiple child rows,
        in the end all parent (main) models should be unique.

        :param result_rows: list of already initialized Models with child models
        populated, each instance is one row in db and some models can duplicate
        :type result_rows: List["Model"]
        :return: list of merged models where each main model is unique
        :rtype: List["Model"]
        """
```

as you can see rtype on DocString is List["Model"] while the *annotated* return type is Sequence["Model"].

If there is any specific reason and I can assist with fixing - I will be very happy to do so!
The package is very helpful and it is much appreciated!

**Comment by collerek** [comment]
It was connected with some kind of error with older version of `mypy`, changed this in last commit and will be merged in next PR #121.
But this actually won't fix your problems as it will still be `List[Model]` not `List[User]` so mypy will still complain.

There is a task for this #112 as typing should be changed to `Genric` but the problem is that `QuerySet` exposed as `objects` is a class variable (populated in metaclass) and looking at https://github.com/python/mypy/issues/2878 it seems like it's not supported.

So probably it will require some structure change to make it work, but if you are willing to help I'm super happy :)

But if you want to do some work fork m2m_fields branch as it has a lot of refactors. I have to finish some minor things before this pr will be merged but it should be later this week.

**Comment by dudil** [comment]
Thanks!

Actually this will solve major part of the issue and will also make the docs consist with the implementation which is alway a good thing :)

I will be happy to assist with #112 - I have an idea about it and will look it in the upcoming days.
Only issue might be with 3.6 support - Generic are much better supported from 3.7+
Is 3.6 still critical to the community?

**Comment by collerek** [comment]
Yeah I would like to maintain the same python version support as pydantic and fastapi, so keep the 3.6. but it can be a fallback to a version that doesn't support proper typing but works in general.

**Comment by dudil** [comment]
NP - I think I was able to solve it.
Let me run some tests and push the PR

**Comment by collerek** [comment]
Don't know if you run the `mypy` on your fix but I got a lot of errors:
![image](https://user-images.githubusercontent.com/16324238/110095038-c21d4380-7d9c-11eb-9f26-d88b79af78a1.png)

Moreover that's not really solving the issue #112, as I did something similar, but that works only when you manually create the QuerySet which you are not supposed to do:
```python
from typing import TYPE_CHECKING

import databases
import pytest
import sqlalchemy

import ormar
from tests.settings import DATABASE_URL

database = databases.Database(DATABASE_URL, force_rollback=True)
metadata = sqlalchemy.MetaData()


class ModelA(ormar.Model):
    class Meta:
        tablename = "audits"
        metadata = metadata
        database = database

    id: int = ormar.Integer(primary_key=True)


@pytest.fixture(autouse=True, scope="module")
def create_test_database():
    engine = sqlalchemy.create_engine(DATABASE_URL)
    metadata.drop_all(engine)
    metadata.create_all(engine)
    yield
    metadata.drop_all(engine)


@pytest.mark.asyncio
async def test_types(cleanup) -> None:
    async with database:
        await ModelA.objects.create()

        qryset = ormar.QuerySet[ModelA](model_cls=ModelA)
        aa = await qryset.get()
        if TYPE_CHECKING:
            reveal_type(aa)

        ab = await ModelA.objects.get()
        if TYPE_CHECKING:
            reveal_type(ab)
```
Produces:
![image](https://user-images.githubusercontent.com/16324238/110096188-f6ddca80-7d9d-11eb-9d69-217d7d397312.png)

The problem is the creation of QuerySet on each model class.

In metaclass.py:
![image](https://user-images.githubusercontent.com/16324238/110096358-1d9c0100-7d9e-11eb-8fd9-4f8f552f4450.png)

And later mypy complains that Model does not have objects property, so it needs to be annotated in Model too:
![image](https://user-images.githubusercontent.com/16324238/110096484-3b696600-7d9e-11eb-91cc-eb7147e832dd.png)

And this above is the problem as you cannot annotate QuerySet[T] in `Model` class as generics are not resolved in class variables.

This can probably be changed to a method `objects()` instead of class variable, but one - it's a breaking change and two- I don't really like the notation of `ModelA.objects().get()` etc. each time.

**Comment by dudil** [comment]
Didn't checked MyPy - sorry for that :(

But I think you are misunderstanding the usage here.
You are not supposed to do QuerySet[T] but rather use it as "it is".
Generic is "smart enough" so it you pass the generic parameter it knows how to bound it all along.
So as long you are consist with that (which you are) - it is working as intended

Given the following code:
''' Python
rows: List[User] = await User.objects.all()
'''

Now works fine (this was my test - I run it with the code examples in the package).
As the return object is User and not just "Model".

**Comment by collerek** [comment]
Yeah but that works only because you annotated the variable with `List[User]`, and the actually returned value from the call is List[any], that's why you can easilly cast it to anything. 

The idea is that the type should be inferred without you needing to annotate the rows variable, as proper type should be returned from all() call.

That's why in my example above `aa` is properly inferred as ModelA, but `ab` is not. Right now if you don't provide your type hint for returned value it's not working. And since whole queryset is typed it should be inferred automatically.

And it's not happening

**Comment by dudil** [comment]
I will try to think with you on the "objects" issue - I agree this is not ideal.

But as for the example - it is also working without the annotation.
And the all() is not returning Any but rather:
'''python
async def all(self, **kwargs: Any) -> Sequence[Optional[TM]]:
'''
I think I am missing something here - let me check the issue with MyPy about the generics - I think there is a solution for that as well

**Comment by collerek** [comment]
It should, but mypy and python cannot properly infer the `objects` class variable and therefore whole `QuerySet` is not inferred properly.

In the end you need to hint `objects` in `Model` class (it's not picked from metaclass) and I have no idea how, as **hinting it just `QuerySet` as it is now breaks the types to QuerySet[Any]** and you cannot use Generic there.

You can annotate it `objects: QuerySet["Model"]` and that way it start to return `Model` not `Any` but basically we are back to square one :)

I even created the issue in so: https://stackoverflow.com/questions/66339371/generic-class-variable-in-python-with-metaclass-and-concrete-base-class , wrote also in mypy gitter but no help so far.

I tried property in metaclass, descriptors, helper functions etc, but with no luck.

**Comment by dudil** [comment]
I have a PR ready which fix most of MyPy issues.
There is one last error which is related to Model itself.

The issue is related to annotation of return of the class inside a class method.
This well known issue in Python which is gradually being fixed (starting with 3.7...)
Have a look into this post - 
https://stackoverflow.com/questions/15853469/putting-current-class-as-return-type-annotation

I didn't changed it and it still raise this last error - it is up to you to decided how exactly to fix that.
Usually in that cases I just return Any and make sure to strict it in the documentation, but if you are brace enough to try the fix propped in the above post - I'll be happy to do it! :-)

**Comment by collerek** [comment]
Yeah, I merged the m2m branch to release it and now we can peacefully work on this one and rest of the issues :) 
Thanks for your work, will check the PR and proposed fix! :)

**Comment by collerek** [comment]
Ok, since this one should be fixed and we still have #112 I will close this one, we can discuss either in 112 or #124
